In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

## II. Exploring data:

#### 1. Center for World University Rankings

Reading csv file

In [2]:
cwur_data = pd.read_csv(r"./data/cwurData.csv")
cwur_data.head()

world_rank                            institution         country  \
0           1                     Harvard University             USA   
1           2  Massachusetts Institute of Technology             USA   
2           3                    Stanford University             USA   
3           4                University of Cambridge  United Kingdom   
4           5     California Institute of Technology             USA   

   national_rank  quality_of_education  alumni_employment  quality_of_faculty  \
0              1                     7                  9                   1   
1              2                     9                 17                   3   
2              3                    17                 11                   5   
3              1                    10                 24                   4   
4              4                     2                 29                   7   

   publications  influence  citations  broad_impact  patents   score  year  
0             1          1          1           NaN        5  100.00  2012  
1            12          4          4           NaN        1   91.67  2012  
2             4          2          2           NaN       15   89.50  2012  
3            16         16         11           NaN       50   86.17  2012  
4            37         22         22           NaN       18   85.21  2012

How many rows and columns are there?

In [3]:
cwur_num_rows, cwur_num_cols = cwur_data.shape
print(f"{cwur_num_rows} rows, {cwur_num_cols} columns")

2200 rows, 14 columns


What is the meaning of each row?

In [4]:
cwur_data.sample(1)

world_rank                   institution country  national_rank  \
1401         202  George Washington University     USA             87   

      quality_of_education  alumni_employment  quality_of_faculty  \
1401                   192                 56                 218   

      publications  influence  citations  broad_impact  patents  score  year  
1401           301        227        264         235.0      326  47.46  2015

- Each row show annual scores, ranks, and some other information of an university (from 2012 to 2015).

Are there any duplicate rows?

In [5]:
cwur_data.duplicated().any()

False

- There are no duplicated rows in this data.

What is the meaning of each column?

In [6]:
list(cwur_data.columns)

['world_rank',
 'institution',
 'country',
 'national_rank',
 'quality_of_education',
 'alumni_employment',
 'quality_of_faculty',
 'publications',
 'influence',
 'citations',
 'broad_impact',
 'patents',
 'score',
 'year']

- According to data description:
    - world_rank: world rank for university
    - institution: name of university
    - country: country of each university
    - national_rank: rank of university within its country
    - quality_of_education: rank for quality of education
    - alumni_employment: rank for alumni employment
    - quality_of_faculty: rank for quality of faculty
    - publications: rank for publications
    - influence: rank for influence
    - citations: number of students at the university
    - broad_impact: rank for broad impact (only available for 2014 and 2015)
    - patents: rank for patents
    - score: total score, used for determining world rank
    - year: year of ranking (2012 to 2015)

What is the current data type of each column? Are
there columns having inappropriate data types?

In [7]:
cwur_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   world_rank            2200 non-null   int64  
 1   institution           2200 non-null   object 
 2   country               2200 non-null   object 
 3   national_rank         2200 non-null   int64  
 4   quality_of_education  2200 non-null   int64  
 5   alumni_employment     2200 non-null   int64  
 6   quality_of_faculty    2200 non-null   int64  
 7   publications          2200 non-null   int64  
 8   influence             2200 non-null   int64  
 9   citations             2200 non-null   int64  
 10  broad_impact          2000 non-null   float64
 11  patents               2200 non-null   int64  
 12  score                 2200 non-null   float64
 13  year                  2200 non-null   int64  
dtypes: float64(2), int64(10), object(2)
memory usage: 240.8+ KB


- Looks like there is no problem with columns' datatype, every column's datatype is accordant with its description

#### 2. Academic Ranking of World Universities (also known as Shanghai Ranking)

Reading csv file

In [8]:
sh_data = pd.read_csv(r"./data/shanghaiData.csv")
sh_data.head()

world_rank                              university_name national_rank  \
0          1                           Harvard University             1   
1          2                      University of Cambridge             1   
2          3                          Stanford University             2   
3          4           University of California, Berkeley             3   
4          5  Massachusetts Institute of Technology (MIT)             4   

   total_score  alumni  award   hici     ns    pub   pcp  year  
0        100.0   100.0  100.0  100.0  100.0  100.0  72.4  2005  
1         73.6    99.8   93.4   53.3   56.6   70.9  66.9  2005  
2         73.4    41.1   72.2   88.5   70.9   72.3  65.0  2005  
3         72.8    71.8   76.0   69.4   73.9   72.2  52.7  2005  
4         70.1    74.0   80.6   66.7   65.8   64.3  53.0  2005

How many rows and columns are there?

In [9]:
sh_num_rows, sh_num_cols = sh_data.shape
print(f"{sh_num_rows} rows, {sh_num_cols} columns")

4897 rows, 11 columns


What is the meaning of each row?

In [10]:
sh_data.sample(1)

world_rank   university_name national_rank  total_score  alumni  award  \
414    401-500  Jilin University             8          NaN     0.0    0.0   

     hici   ns   pub  pcp  year  
414   0.0  6.2  32.3  9.1  2005

- Each row show annual scores, ranks (or rank ranges), and some other information of an university (from 2005 to 2015)

Are there any duplicate rows?

In [11]:
sh_data.duplicated().any()

False

- There are also no duplicated rows in this data.

What is the meaning of each column?

In [12]:
list(sh_data.columns)

['world_rank',
 'university_name',
 'national_rank',
 'total_score',
 'alumni',
 'award',
 'hici',
 'ns',
 'pub',
 'pcp',
 'year']

- According to data description:
    - world_rank: world rank for university. Contains rank ranges and equal ranks (eg. 101-152)
    - university_name: name of university
    - national_rank: rank of university within its country
    - total_score: total score, used to determine rank
    - alumni: Alumni Score, based on the number of alumni of an institution winning nobel prizes and fields medals
    - award: Award Score, based on the number of staff of an institution winning Nobel Prizes in Physics, Chemistry, Medicine, and Economics and Fields Medals in Mathematics
    - hici: HiCi Score, based on the number of Highly Cited Researchers selected by Thomson Reuters
    - ns: N&S Score, based on the number of papers published in Nature and Science
    - pub: PUB Score, based on total number of papers indexed in the Science Citation Index-Expanded and Social Science Citation Index
    - pcp: PCP Score, the weighted scores of the above five indicators divided by the number of full time academic staff
    - year: year of ranking (2005 to 2015)

What is the current data type of each column? Are
there columns having inappropriate data types?

In [13]:
sh_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4897 entries, 0 to 4896
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   world_rank       4897 non-null   object 
 1   university_name  4896 non-null   object 
 2   national_rank    4896 non-null   object 
 3   total_score      1101 non-null   float64
 4   alumni           4896 non-null   float64
 5   award            4895 non-null   float64
 6   hici             4895 non-null   float64
 7   ns               4875 non-null   float64
 8   pub              4895 non-null   float64
 9   pcp              4895 non-null   float64
 10  year             4897 non-null   int64  
dtypes: float64(7), int64(1), object(3)
memory usage: 421.0+ KB


- Same as cwur_data, Looks like there is also no problem with columns' datatype, every column's datatype is accordant with its description

#### 3. Schools (universities) and countries data

Reading csv file

In [14]:
school_and_country = pd.read_csv(r"./data/school_and_country_table.csv")

How many rows and columns are there?

In [15]:
sac_num_rows, sac_num_cols = school_and_country.shape
print(f"{sac_num_rows} rows, {sac_num_cols} columns")

818 rows, 2 columns


What is the meaning of each row?

In [16]:
school_and_country.sample(1)

school_name  country
511  Graz University of Technology  Austria

- Each row show a university name and the country where it is located

What is the meaning of each column?

In [17]:
list(school_and_country.columns)

['school_name', 'country']

- school_name: the name of the university
- country: the country where the university is located

What is the current data type of each column? Are
there columns having inappropriate data types?

In [18]:
school_and_country.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 818 entries, 0 to 817
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   school_name  818 non-null    object
 1   country      818 non-null    object
dtypes: object(2)
memory usage: 12.9+ KB


- All columns' datatype looks fine.

#### 4. Educational expenditure supplementary data

Reading csv file

In [19]:
education_expenditure_data = pd.read_csv(r"./data/education_expenditure_supplementary_data.csv", lineterminator='\r')

How many rows and columns are there?

In [20]:
eed_num_rows, eed_num_cols = education_expenditure_data.shape
print(f"{eed_num_rows} rows, {eed_num_cols} columns")

333 rows, 9 columns


What is the meaning of each row?

In [21]:
education_expenditure_data.sample(1)

country                          institute_type direct_expenditure_type  \
72    Brazil  Elementary and Secondary Institutions                   Public   

    1995  2000  2005  2009  2010  2011  
72   NaN   NaN   3.3   4.3   4.3   4.4

- Each row show expenditure on education of country over years (1995 to 2011), it is represented as a percentage of gross domestic product (% GDP) 

What is the meaning of each column?

In [22]:
list(education_expenditure_data.columns)

['country',
 'institute_type',
 'direct_expenditure_type',
 '1995',
 '2000',
 '2005',
 '2009',
 '2010',
 '2011']

- According to data description:
    - country: A set of OECD countries, plus Brazil and the Russian Federation, plus an OECD average
    - institute_type: All Institutes (including preprimary education and subsidies to households, not separately shown , Elementary and secondary institutions (excludes preprimary), Higher education institutions
    - direct_expenditure_type: Public direct expenditure, private direct expenditure, or total (public + private) direct expenditure.
    - 1995, 2000, 2005, 2009, 2010, 2011: The % GDP of expenditure on education of the respective year

What is the current data type of each column? Are
there columns having inappropriate data types?

In [23]:
education_expenditure_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 333 entries, 0 to 332
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   country                  333 non-null    object 
 1   institute_type           333 non-null    object 
 2   direct_expenditure_type  333 non-null    object 
 3   1995                     94 non-null     float64
 4   2000                     99 non-null     float64
 5   2005                     106 non-null    float64
 6   2009                     103 non-null    float64
 7   2010                     101 non-null    float64
 8   2011                     282 non-null    float64
dtypes: float64(6), object(3)
memory usage: 23.5+ KB


There is no problem with the columns' datatype in this data. All match its description

## III. Asking meaningful questions:

#### 1. What are top 5 countries that have most universities in top 30 ranking list, over years (from 2012 to 2015), according to Center for World University Rankings?

- The answer to this question gives a general look in ranks of top countries' universities. In detail:
    - It will reflect some top education systems in the world.
    - Some students who intend to study abroad should first look for the country they want to study, they can look at this answer as an additional criteria to see whether it is suitable for their thoroughly-known language, or some other criteria...

#### 2. According to Shanghai Ranking, how does public spending on higher education affect the average total score of each country in 2011?

- To answer this question, we should answer 2 sub questions and make an observation on the result:
    1. What are top 5 countries with highest average score and their public expenditure on higher education?
    2. What are top 5 countries in public expenditure on higher education and their average score?

- The answer to this question gives a general look in relationship between rankings and education expenditure. There might be some useful information to answer other questions such as:
    - Is there a relationship between expenditure on education and the economic growth?
    - How should government spend on their country's educational system?
    - Does spending more on education lead to better university rankings?

## IV. Preprocessing + analyzing data to answer each question:

#### 1. What are top 5 countries that have most universities in top 30 ranking list, over years (from 2012 to 2015), according to Center for World University Rankings?

First look at the data's info again and see if there is any missing data.

In [24]:
cwur_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   world_rank            2200 non-null   int64  
 1   institution           2200 non-null   object 
 2   country               2200 non-null   object 
 3   national_rank         2200 non-null   int64  
 4   quality_of_education  2200 non-null   int64  
 5   alumni_employment     2200 non-null   int64  
 6   quality_of_faculty    2200 non-null   int64  
 7   publications          2200 non-null   int64  
 8   influence             2200 non-null   int64  
 9   citations             2200 non-null   int64  
 10  broad_impact          2000 non-null   float64
 11  patents               2200 non-null   int64  
 12  score                 2200 non-null   float64
 13  year                  2200 non-null   int64  
dtypes: float64(2), int64(10), object(2)
memory usage: 240.8+ KB


- The ***broad_impact*** column seem to have some missing values, but it will not be matter for this question, so we can leave it for now. The remaining columns look ok.

Next, we filter out the top 30 ranking of all data frame:

In [25]:
df = cwur_data[cwur_data['world_rank'] <= 30]
df

world_rank                              institution         country  \
0              1                       Harvard University             USA   
1              2    Massachusetts Institute of Technology             USA   
2              3                      Stanford University             USA   
3              4                  University of Cambridge  United Kingdom   
4              5       California Institute of Technology             USA   
...          ...                                      ...             ...   
1225          26  University of California, San Francisco             USA   
1226          27                University College London  United Kingdom   
1227          28                          Duke University             USA   
1228          29                   Rockefeller University             USA   
1229          30            University of Texas at Austin             USA   

      national_rank  quality_of_education  alumni_employment  \
0                 1                     7                  9   
1                 2                     9                 17   
2                 3                    17                 11   
3                 1                    10                 24   
4                 4                     2                 29   
...             ...                   ...                ...   
1225             19                   367                567   
1226              3                    21                447   
1227             20                    74                 31   
1228             21                    49                567   
1229             22                    75                 36   

      quality_of_faculty  publications  influence  citations  broad_impact  \
0                      1             1          1          1           NaN   
1                      3            12          4          4           NaN   
2                      5             4          2          2           NaN   
3                      4            16         16         11           NaN   
4                      7            37         22         22           NaN   
...                  ...           ...        ...        ...           ...   
1225                  20            22          7         15          11.0   
1226                  27            12         22         15          18.0   
1227                  56            19         23         15          16.0   
1228                  12           372         28        115          70.0   
1229                  25            48         50         26          71.0   

      patents   score  year  
0           5  100.00  2012  
1           1   91.67  2012  
2          15   89.50  2012  
3          50   86.17  2012  
4          18   85.21  2012  
...       ...     ...   ...  
1225       56   63.69  2015  
1226       67   62.27  2015  
1227       33   61.55  2015  
1228      289   61.28  2015  
1229       38   61.14  2015  

[120 rows x 14 columns]

We group by ***year*** and ***country***, and then apply row counting on the ***institution*** to get the number of universities of each country, in annual rankings.

In [26]:
df_grouped = df.groupby(["year", "country"])["institution"].size()
df_grouped.to_frame() # for prettier look

institution
year country                    
2012 Israel                    2
     Japan                     2
     Switzerland               1
     USA                      22
     United Kingdom            3
2013 Canada                    1
     Israel                    1
     Japan                     2
     Switzerland               1
     USA                      21
     United Kingdom            4
2014 Israel                    1
     Japan                     2
     South Korea               1
     Switzerland               1
     USA                      22
     United Kingdom            3
2015 Israel                    1
     Japan                     2
     South Korea               1
     Switzerland               1
     USA                      22
     United Kingdom            3

Finally, we get top 5 countries that have largest number of institutions. To do this, we need to apply group by method again by year and filter out the 5 largest institutions, but this time in group by method, ***group_keys*** parameter should be False to avoid repeating the ***year*** column.

In [66]:
answer_1 = df_grouped.groupby("year", group_keys=False).nlargest(5)
answer_1.to_frame() # for prettier look

institution
year country                    
2012 USA                      22
     United Kingdom            3
     Israel                    2
     Japan                     2
     Switzerland               1
2013 USA                      21
     United Kingdom            4
     Japan                     2
     Canada                    1
     Israel                    1
2014 USA                      22
     United Kingdom            3
     Japan                     2
     Israel                    1
     South Korea               1
2015 USA                      22
     United Kingdom            3
     Japan                     2
     Israel                    1
     South Korea               1

##### Result overview:

- Looking at the result above, we can see that over years (from 2012 to 2015), the United States of America is likely dominating other countries in the Center for World University Rankings itself. The USA's national language is English - a world popular language, so this is usually the prior choice of international students.

#### 2. According to Shanghai Ranking, how does public spending on higher education affect the average total score of each country in 2011?

##### Preprocessing data:

First look at the data's info again and see if there is any missing data.

In [28]:
sh_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4897 entries, 0 to 4896
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   world_rank       4897 non-null   object 
 1   university_name  4896 non-null   object 
 2   national_rank    4896 non-null   object 
 3   total_score      1101 non-null   float64
 4   alumni           4896 non-null   float64
 5   award            4895 non-null   float64
 6   hici             4895 non-null   float64
 7   ns               4875 non-null   float64
 8   pub              4895 non-null   float64
 9   pcp              4895 non-null   float64
 10  year             4897 non-null   int64  
dtypes: float64(7), int64(1), object(3)
memory usage: 421.0+ KB


- There are 4897 rows but only 4896 non-null in ***university_name*** column, there could be 1 blank row in the data, we can check this:

In [29]:
sh_data[sh_data['university_name'].isna()]

world_rank university_name national_rank  total_score  alumni  award  \
3896         99             NaN           NaN          NaN     NaN    NaN   

      hici  ns  pub  pcp  year  
3896   NaN NaN  NaN  NaN  2013

- This row has no meaning, so we can safely drop it (i.e filter out all the rows that ***university_name*** is not null).

In [30]:
df = sh_data[sh_data['university_name'].notna()]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4896 entries, 0 to 4896
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   world_rank       4896 non-null   object 
 1   university_name  4896 non-null   object 
 2   national_rank    4896 non-null   object 
 3   total_score      1101 non-null   float64
 4   alumni           4896 non-null   float64
 5   award            4895 non-null   float64
 6   hici             4895 non-null   float64
 7   ns               4875 non-null   float64
 8   pub              4895 non-null   float64
 9   pcp              4895 non-null   float64
 10  year             4896 non-null   int64  
dtypes: float64(7), int64(1), object(3)
memory usage: 459.0+ KB


We just need the data in the year of 2011, so we can filter out this first to speed up the process.

In [31]:
df = df[df['year']==2011]
df

world_rank                              university_name national_rank  \
3014          1                           Harvard University             1   
3015          2                          Stanford University             2   
3016          3  Massachusetts Institute of Technology (MIT)             3   
3017          4           University of California, Berkeley             4   
3018          5                      University of Cambridge             1   
...         ...                                          ...           ...   
3509    401-500                             Wuhan University         13-23   
3510    401-500                            Xiamen University         13-23   
3511    401-500                    Xian Jiao Tong University         13-23   
3512    401-500                         Yamaguchi University         17-23   
3513    401-500                              York University         19-22   

      total_score  alumni  award   hici     ns    pub   pcp  year  
3014        100.0   100.0  100.0  100.0  100.0  100.0  70.5  2011  
3015         72.6    41.2   78.4   88.4   70.2   70.3  48.6  2011  
3016         72.0    72.8   81.9   67.9   70.6   60.6  63.7  2011  
3017         71.9    68.3   79.3   70.0   69.5   69.4  53.1  2011  
3018         70.0    87.1   96.7   54.5   54.1   65.1  52.0  2011  
...           ...     ...    ...    ...    ...    ...   ...   ...  
3509          NaN     0.0    0.0    0.0    2.6   36.6  10.0  2011  
3510          NaN     0.0    0.0    0.0    8.9   32.9  11.6  2011  
3511          NaN     0.0    0.0    0.0    5.0   39.2  14.2  2011  
3512          NaN     0.0    0.0   16.1    2.8   20.0  13.2  2011  
3513          NaN     0.0    0.0    0.0   11.0   31.5  11.7  2011  

[500 rows x 11 columns]

There are a lot of missing values of ***total_score*** column, this can affect the result of this question. Fortunately, we can calculate total score based on the other scores, the formula is:

$total\_score = 0.1 \times alumni + 0.2 \times award + 0.2 \times hici + 0.2 \times ns + 0.2 \times pub + 0.1 \times pcp$

In [32]:
df['total_score'] = 0.1 * (df['alumni'] + df['pcp']) \
              + 0.2 * (df['award'] + df['hici'] + df['ns'] + df['pub'])
df

world_rank                              university_name national_rank  \
3014          1                           Harvard University             1   
3015          2                          Stanford University             2   
3016          3  Massachusetts Institute of Technology (MIT)             3   
3017          4           University of California, Berkeley             4   
3018          5                      University of Cambridge             1   
...         ...                                          ...           ...   
3509    401-500                             Wuhan University         13-23   
3510    401-500                            Xiamen University         13-23   
3511    401-500                    Xian Jiao Tong University         13-23   
3512    401-500                         Yamaguchi University         17-23   
3513    401-500                              York University         19-22   

      total_score  alumni  award   hici     ns    pub   pcp  year  
3014        97.05   100.0  100.0  100.0  100.0  100.0  70.5  2011  
3015        70.44    41.2   78.4   88.4   70.2   70.3  48.6  2011  
3016        69.85    72.8   81.9   67.9   70.6   60.6  63.7  2011  
3017        69.78    68.3   79.3   70.0   69.5   69.4  53.1  2011  
3018        67.99    87.1   96.7   54.5   54.1   65.1  52.0  2011  
...           ...     ...    ...    ...    ...    ...   ...   ...  
3509         8.84     0.0    0.0    0.0    2.6   36.6  10.0  2011  
3510         9.52     0.0    0.0    0.0    8.9   32.9  11.6  2011  
3511        10.26     0.0    0.0    0.0    5.0   39.2  14.2  2011  
3512         9.10     0.0    0.0   16.1    2.8   20.0  13.2  2011  
3513         9.67     0.0    0.0    0.0   11.0   31.5  11.7  2011  

[500 rows x 11 columns]

Lets see if there is still any NaN value in the ***total_score*** column after calculation.

In [33]:
df[df['total_score'].isna()].shape[0]

3

Looks like there are still 3 NaN values because one of the score is missing (NaN), so we can drop this since we have no way to fix.

In [34]:
df = df[df['total_score'].notna()]
df

world_rank                              university_name national_rank  \
3014          1                           Harvard University             1   
3015          2                          Stanford University             2   
3016          3  Massachusetts Institute of Technology (MIT)             3   
3017          4           University of California, Berkeley             4   
3018          5                      University of Cambridge             1   
...         ...                                          ...           ...   
3509    401-500                             Wuhan University         13-23   
3510    401-500                            Xiamen University         13-23   
3511    401-500                    Xian Jiao Tong University         13-23   
3512    401-500                         Yamaguchi University         17-23   
3513    401-500                              York University         19-22   

      total_score  alumni  award   hici     ns    pub   pcp  year  
3014        97.05   100.0  100.0  100.0  100.0  100.0  70.5  2011  
3015        70.44    41.2   78.4   88.4   70.2   70.3  48.6  2011  
3016        69.85    72.8   81.9   67.9   70.6   60.6  63.7  2011  
3017        69.78    68.3   79.3   70.0   69.5   69.4  53.1  2011  
3018        67.99    87.1   96.7   54.5   54.1   65.1  52.0  2011  
...           ...     ...    ...    ...    ...    ...   ...   ...  
3509         8.84     0.0    0.0    0.0    2.6   36.6  10.0  2011  
3510         9.52     0.0    0.0    0.0    8.9   32.9  11.6  2011  
3511        10.26     0.0    0.0    0.0    5.0   39.2  14.2  2011  
3512         9.10     0.0    0.0   16.1    2.8   20.0  13.2  2011  
3513         9.67     0.0    0.0    0.0   11.0   31.5  11.7  2011  

[497 rows x 11 columns]

Now to answer the question, we are still missing a column ***country*** in Shanghai data frame, we need to preprocess this.

First focus on the ***university_name*** column, how many unique values are there?

In [35]:
unique_universities = pd.DataFrame(df['university_name'].unique(), columns=['university_name'])
unique_universities

university_name
0                             Harvard University
1                            Stanford University
2    Massachusetts Institute of Technology (MIT)
3             University of California, Berkeley
4                        University of Cambridge
..                                           ...
492                             Wuhan University
493                            Xiamen University
494                    Xian Jiao Tong University
495                         Yamaguchi University
496                              York University

[497 rows x 1 columns]

- We want to get unique values because it will be more optimal in later process (matching universities and countries)

Next, we will look at the school_and_country data frame again

In [36]:
school_and_country

school_name                   country
0                       Harvard University  United States of America
1       California Institute of Technology  United States of America
2    Massachusetts Institute of Technology  United States of America
3                      Stanford University  United States of America
4                     Princeton University  United States of America
..                                     ...                       ...
813                      Xidian University                     China
814                    Yeungnam University               South Korea
815            Yıldız Technical University                    Turkey
816               Yokohama City University                     Japan
817           Yokohama National University                     Japan

[818 rows x 2 columns]

Are there any duplicated rows in school_and_country data frame? If yes, we should first make them unique.

In [37]:
school_and_country.duplicated().any()

False

- There is no duplicated rows in the data, so we can safely move on

There is a small inconsistent between the name of universities column, so we will rename it.

In [38]:
school_and_country = school_and_country.rename(columns={'school_name':'university_name'})
school_and_country

university_name                   country
0                       Harvard University  United States of America
1       California Institute of Technology  United States of America
2    Massachusetts Institute of Technology  United States of America
3                      Stanford University  United States of America
4                     Princeton University  United States of America
..                                     ...                       ...
813                      Xidian University                     China
814                    Yeungnam University               South Korea
815            Yıldız Technical University                    Turkey
816               Yokohama City University                     Japan
817           Yokohama National University                     Japan

[818 rows x 2 columns]

Create a global university and country dictionary, after each process finish and a new data frame is created, add all of non-null (processed) value into the dictionary.

In [39]:
my_school_and_country_dict = {}

Now we can merge the unique_universities with the school_country data above to get the country of each university, the ***how*** parameter show be "left" because there may be universities not in school_and_country data, we want this to be NaN.

In [40]:
df_merged = unique_universities.merge(school_and_country, how="left")

# store the filled data into dictionary
my_school_and_country_dict.update( \
    df_merged[df_merged['country'].notna()].set_index('university_name').to_dict()['country'])
    
df_merged

university_name                   country
0                             Harvard University  United States of America
1                            Stanford University  United States of America
2    Massachusetts Institute of Technology (MIT)                       NaN
3             University of California, Berkeley  United States of America
4                        University of Cambridge            United Kingdom
..                                           ...                       ...
492                             Wuhan University                     China
493                            Xiamen University                     China
494                    Xian Jiao Tong University                       NaN
495                         Yamaguchi University                       NaN
496                              York University                    Canada

[497 rows x 2 columns]

How many universities do not have their country after merging?

In [41]:
df_merged[df_merged['country'].isna()]

university_name country
2    Massachusetts Institute of Technology (MIT)     NaN
16       University of California, San Francisco     NaN
17                  The Johns Hopkins University     NaN
18             University of Wisconsin - Madison     NaN
20                       The University of Tokyo     NaN
..                                           ...     ...
484          University of Wisconsin - Milwaukee     NaN
487                         University of Zagreb     NaN
491                     Vrije University Brussel     NaN
494                    Xian Jiao Tong University     NaN
495                         Yamaguchi University     NaN

[192 rows x 2 columns]

- There are still 192 universities of which country is missing. If we look for these university in the school_and_country data once again, we can see most of these are in different orders, or there are some additional words in the name.

Now, there are a lot of modules to help us fix this (e.g Regex), but after testing, we found out using module called ***fuzzywuzzy*** is very effective, it is a Python module that supports matching strings.

In [42]:
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

We create a dictionary to store universities and its corresponding countries according to school_and_country data

In [43]:
school_and_country_dict = school_and_country.set_index('university_name').to_dict()['country']
schools = school_and_country_dict.keys()

We compare each university name in the merged data with each university in school_and_country data. If the score is over 80 (calculated by fuzzywuzzy), we store the country corresponding to that university.

In [44]:
def compare(s, min_score):
    university_name, score = process.extractOne(s, schools, scorer=fuzz.token_set_ratio)
    if score >= min_score:
        return school_and_country_dict[university_name]

    return np.nan

In [45]:
df_fuzzywuzzy = df_merged[df_merged['country'].isna()].copy()
df_fuzzywuzzy['country'] = df_fuzzywuzzy['university_name'].apply(lambda s: compare(str(s), 80))

# store the filled data into dictionary
my_school_and_country_dict.update( \
    df_fuzzywuzzy[df_fuzzywuzzy['country'].notna()].set_index('university_name').to_dict()['country'])

df_fuzzywuzzy

university_name                   country
2    Massachusetts Institute of Technology (MIT)  United States of America
16       University of California, San Francisco  United States of America
17                  The Johns Hopkins University  United States of America
18             University of Wisconsin - Madison  United States of America
20                       The University of Tokyo                     Japan
..                                           ...                       ...
484          University of Wisconsin - Milwaukee  United States of America
487                         University of Zagreb                    Turkey
491                     Vrije University Brussel                   Belgium
494                    Xian Jiao Tong University                     China
495                         Yamaguchi University                    Turkey

[192 rows x 2 columns]

Do we still have any unfilled values?

In [46]:
df_fuzzywuzzy[df_fuzzywuzzy['country'].isna()]

university_name country
102                         Baylor College of Medicine     NaN
109                                Mayo Medical School     NaN
161                     Mount Sinai School of Medicine     NaN
321  Industrial Physics and Chemistry Higher Educat...     NaN
411            Ecole National Superieure Mines - Paris     NaN
416                            Hannover Medical School     NaN
426     London School of Hygiene and Tropical Medicine     NaN

- There are still 7 NaN values, because these university name does not exist in school_and_country data, so how can we fill these? The number of values is not too large, so we can do it manually by asking Google.

In [47]:
df_processed = df_fuzzywuzzy[df_fuzzywuzzy['country'].isna()].copy()

def country_fill(name, country):
    df_processed.loc[df_processed['university_name'] == name, ['country']] = country

country_fill("Baylor College of Medicine", "United States of America")
country_fill("Mayo Medical School", "United States of America")
country_fill("Mount Sinai School of Medicine", "United States of America")
country_fill("Ecole National Superieure Mines - Paris", "France")
country_fill("Industrial Physics and Chemistry Higher Educational Institution - Paris", "France")
country_fill("London School of Hygiene and Tropical Medicine", "United Kingdom")
country_fill("Hannover Medical School", "Germany")

# store the filled data into dictionary
my_school_and_country_dict.update( \
    df_processed.set_index('university_name').to_dict()['country'])

df_processed

university_name  \
102                         Baylor College of Medicine   
109                                Mayo Medical School   
161                     Mount Sinai School of Medicine   
321  Industrial Physics and Chemistry Higher Educat...   
411            Ecole National Superieure Mines - Paris   
416                            Hannover Medical School   
426     London School of Hygiene and Tropical Medicine   

                      country  
102  United States of America  
109  United States of America  
161  United States of America  
321                    France  
411                    France  
416                   Germany  
426            United Kingdom

After getting all of the countries ready (stored in my_school_and_country_dict), we fill it in a new ***country*** column, we should do this with a temporary data frame, or else it will raise "SettingWithCopyWarning"

In [48]:
df.reset_index(drop=True,inplace=True)
tmp = df.copy()
tmp['country'] = df['university_name'].apply(lambda s: my_school_and_country_dict[s])
df = tmp
df

world_rank                              university_name national_rank  \
0            1                           Harvard University             1   
1            2                          Stanford University             2   
2            3  Massachusetts Institute of Technology (MIT)             3   
3            4           University of California, Berkeley             4   
4            5                      University of Cambridge             1   
..         ...                                          ...           ...   
492    401-500                             Wuhan University         13-23   
493    401-500                            Xiamen University         13-23   
494    401-500                    Xian Jiao Tong University         13-23   
495    401-500                         Yamaguchi University         17-23   
496    401-500                              York University         19-22   

     total_score  alumni  award   hici     ns    pub   pcp  year  \
0          97.05   100.0  100.0  100.0  100.0  100.0  70.5  2011   
1          70.44    41.2   78.4   88.4   70.2   70.3  48.6  2011   
2          69.85    72.8   81.9   67.9   70.6   60.6  63.7  2011   
3          69.78    68.3   79.3   70.0   69.5   69.4  53.1  2011   
4          67.99    87.1   96.7   54.5   54.1   65.1  52.0  2011   
..           ...     ...    ...    ...    ...    ...   ...   ...   
492         8.84     0.0    0.0    0.0    2.6   36.6  10.0  2011   
493         9.52     0.0    0.0    0.0    8.9   32.9  11.6  2011   
494        10.26     0.0    0.0    0.0    5.0   39.2  14.2  2011   
495         9.10     0.0    0.0   16.1    2.8   20.0  13.2  2011   
496         9.67     0.0    0.0    0.0   11.0   31.5  11.7  2011   

                      country  
0    United States of America  
1    United States of America  
2    United States of America  
3    United States of America  
4              United Kingdom  
..                        ...  
492                     China  
493                     China  
494                     China  
495                    Turkey  
496                    Canada  

[497 rows x 12 columns]

Just to make sure that no NaN country

In [49]:
df[df['country'].isna()]

Empty DataFrame
Columns: [world_rank, university_name, national_rank, total_score, alumni, award, hici, ns, pub, pcp, year, country]
Index: []

In [50]:
df

world_rank                              university_name national_rank  \
0            1                           Harvard University             1   
1            2                          Stanford University             2   
2            3  Massachusetts Institute of Technology (MIT)             3   
3            4           University of California, Berkeley             4   
4            5                      University of Cambridge             1   
..         ...                                          ...           ...   
492    401-500                             Wuhan University         13-23   
493    401-500                            Xiamen University         13-23   
494    401-500                    Xian Jiao Tong University         13-23   
495    401-500                         Yamaguchi University         17-23   
496    401-500                              York University         19-22   

     total_score  alumni  award   hici     ns    pub   pcp  year  \
0          97.05   100.0  100.0  100.0  100.0  100.0  70.5  2011   
1          70.44    41.2   78.4   88.4   70.2   70.3  48.6  2011   
2          69.85    72.8   81.9   67.9   70.6   60.6  63.7  2011   
3          69.78    68.3   79.3   70.0   69.5   69.4  53.1  2011   
4          67.99    87.1   96.7   54.5   54.1   65.1  52.0  2011   
..           ...     ...    ...    ...    ...    ...   ...   ...   
492         8.84     0.0    0.0    0.0    2.6   36.6  10.0  2011   
493         9.52     0.0    0.0    0.0    8.9   32.9  11.6  2011   
494        10.26     0.0    0.0    0.0    5.0   39.2  14.2  2011   
495         9.10     0.0    0.0   16.1    2.8   20.0  13.2  2011   
496         9.67     0.0    0.0    0.0   11.0   31.5  11.7  2011   

                      country  
0    United States of America  
1    United States of America  
2    United States of America  
3    United States of America  
4              United Kingdom  
..                        ...  
492                     China  
493                     China  
494                     China  
495                    Turkey  
496                    Canada  

[497 rows x 12 columns]

So we now finish with all of the preprocessing, we can move on and answer the questions.

##### Answering the sub questions:

First, we will construct the top 5 score table, we group by the country and get the mean total_score of each country, also reset the index to make it a data frame.

In [69]:
df_scores = df.groupby('country')['total_score'].mean().reset_index() 
df_scores.head()

country  total_score
0  Australia    15.774737
1    Austria    12.708750
2    Belgium    16.100000
3     Brazil    12.602857
4     Canada    17.365000

Then we get the top 5 countries with largest average score.

In [52]:
df_top_5_scores = df_scores.nlargest(5, 'total_score')
df_top_5_scores = df_top_5_scores.rename(columns={'total_score':'average_total_score'}) # for prettier look
df_top_5_scores

country  average_total_score
36               Switzerland            25.224286
40  United States of America            23.633219
8                    Denmark            22.415000
39            United Kingdom            21.337027
22               Netherlands            19.626667

Now construct the top 5 expenditure on education.

In [53]:
education_expenditure_data

country                  institute_type  \
0            OECD Average               All Institutions    
1               Australia               All Institutions    
2                 Austria               All Institutions    
3                 Belgium               All Institutions    
4                  Canada               All Institutions    
..                    ...                             ...   
328                Turkey  Higher Education Institutions    
329        United Kingdom  Higher Education Institutions    
330         United States  Higher Education Institutions    
331                Brazil  Higher Education Institutions    
332    Russian Federation  Higher Education Institutions    

    direct_expenditure_type  1995  2000  2005  2009  2010  2011  
0                    Public   4.9   4.9   5.0   5.4   5.4   5.3  
1                    Public   4.5   4.6   4.3   4.5   4.6   4.3  
2                    Public   5.3   5.4   5.2   5.7   5.6   5.5  
3                    Public   5.0   5.1   5.8   6.4   6.4   6.4  
4                    Public   5.8   5.2   4.8   5.0   5.2   NaN  
..                      ...   ...   ...   ...   ...   ...   ...  
328                   Total   NaN   NaN   NaN   NaN   NaN   NaN  
329                   Total   NaN   NaN   NaN   NaN   NaN   1.2  
330                   Total   NaN   NaN   NaN   NaN   NaN   2.7  
331                   Total   NaN   NaN   NaN   NaN   NaN   NaN  
332                   Total   NaN   NaN   NaN   NaN   NaN   1.4  

[333 rows x 9 columns]

When doing this, we notice a small problem with ***institute_type*** column

In [54]:
for element in education_expenditure_data['institute_type'].unique():
    print(repr(element))

'All Institutions '
'Elementary and Secondary Institutions '
'Higher Education Institutions '


There are some redundant characters at the end of the string, we should strip it in prior.

In [55]:
education_expenditure_data['institute_type'] = education_expenditure_data['institute_type'].apply(lambda s: s.strip())
for element in education_expenditure_data['institute_type'].unique():
    print(repr(element))

'All Institutions'
'Elementary and Secondary Institutions'
'Higher Education Institutions'


Now we filter out the rows that have both *Public* and *Higher Education Institutions*.

In [56]:
df_expenditure = education_expenditure_data.copy()
df_expenditure = df_expenditure[(df_expenditure['direct_expenditure_type'] == 'Public') \
                         & (df_expenditure['institute_type'] == 'Higher Education Institutions')]
df_expenditure.head()

country                 institute_type direct_expenditure_type  1995  \
74  OECD Average  Higher Education Institutions                  Public   0.9   
75     Australia  Higher Education Institutions                  Public   1.2   
76       Austria  Higher Education Institutions                  Public   0.9   
77       Belgium  Higher Education Institutions                  Public   0.9   
78        Canada  Higher Education Institutions                  Public   1.5   

    2000  2005  2009  2010  2011  
74   1.0   1.0   1.1   1.1   1.1  
75   0.8   0.8   0.7   0.8   0.7  
76   1.2   1.2   1.4   1.5   1.4  
77   1.2   1.2   1.4   1.4   1.3  
78   1.6   1.5   1.5   1.6   NaN

The question needs data in year 2011, but some are missing, we can estimate this using data from previous years (front fill).

In [57]:
df_expenditure['2011'] = education_expenditure_data.iloc[:, 3:9].fillna(axis=1, method='ffill')['2011']
df_expenditure.head()

country                 institute_type direct_expenditure_type  1995  \
74  OECD Average  Higher Education Institutions                  Public   0.9   
75     Australia  Higher Education Institutions                  Public   1.2   
76       Austria  Higher Education Institutions                  Public   0.9   
77       Belgium  Higher Education Institutions                  Public   0.9   
78        Canada  Higher Education Institutions                  Public   1.5   

    2000  2005  2009  2010  2011  
74   1.0   1.0   1.1   1.1   1.1  
75   0.8   0.8   0.7   0.8   0.7  
76   1.2   1.2   1.4   1.5   1.4  
77   1.2   1.2   1.4   1.4   1.3  
78   1.6   1.5   1.5   1.6   1.6

We still have some NaN values in 2011, this means no value of previous year was found, so we can drop all of these rows

In [58]:
df_expenditure = df_expenditure[df_expenditure['2011'].notna()]
df_expenditure = df_expenditure.loc[:, ['country', '2011']] # for easier look
df_expenditure.head()

country  2011
74  OECD Average   1.1
75     Australia   0.7
76       Austria   1.4
77       Belgium   1.3
78        Canada   1.6

Now we can get the top 5 countries that have the largest public expenditure on higher education.

In [71]:
df_top_5_expenditure = df_expenditure.nlargest(5, ['2011']).reset_index(drop=True) # for prettier look
df_top_5_expenditure

country  2011
0  Finland   1.9
1  Denmark   1.8
2   Canada   1.6
3   Norway   1.6
4   Sweden   1.6

##### Sub question 1: What are top 5 countries with highest average score and their public expenditure on higher education?

We then merge the *top 5 scores* with *expenditure data* frame to get the education expenditure of top 5 highest score countries.

In [60]:
df_top_5_scores.merge(df_expenditure)

country  average_total_score  2011
0     Switzerland            25.224286   1.3
1         Denmark            22.415000   1.8
2  United Kingdom            21.337027   0.9
3     Netherlands            19.626667   1.3

Wait, where is the United States of America?, if we look at the education_expenditure again:

In [61]:
df_expenditure.tail(3)

country  2011
108         United States   0.9
109                Brazil   0.9
110    Russian Federation   0.9

The United States of America in the top 5 scores data is actually the United States in education expenditure data. Fortunately, there are no other exception so we can fix this just by replacing string.

In [62]:
df_top_5_scores['country'] = df_top_5_scores['country'].replace('United States of America', 'United States')
df_top_5_scores

country  average_total_score
36     Switzerland            25.224286
40   United States            23.633219
8          Denmark            22.415000
39  United Kingdom            21.337027
22     Netherlands            19.626667

We now can merge the *top 5 scores* with *expenditure* data frame to get the education expenditure of top 5 highest score countries.

In [63]:
df_top_5_scores = df_top_5_scores.merge(df_expenditure)
df_top_5_scores

country  average_total_score  2011
0     Switzerland            25.224286   1.3
1   United States            23.633219   0.9
2         Denmark            22.415000   1.8
3  United Kingdom            21.337027   0.9
4     Netherlands            19.626667   1.3

##### Sub question 2: What are top 5 countries in public expenditure on higher education and their average score?

To get the top 5 in education expenditure countries' score, we merge the *top 5 expenditure* data with *scores* data. 

In [64]:
df_top_5_expenditure = df_top_5_expenditure.merge(df_scores)
df_top_5_expenditure.rename(columns={'total_score':'average_total_score'}, inplace=True) # for tables consistency
df_top_5_expenditure

country  2011  average_total_score
0  Finland   1.9            14.120000
1  Denmark   1.8            22.415000
2   Canada   1.6            17.365000
3   Norway   1.6            16.450000
4   Sweden   1.6            18.320909

##### A general observation:

- From two tables above, we can see that:
    - Country which is in top 5 average score does not spend a lot on their higher education system (if we compare with the top 5 expenditure on higher education ones).
    - In contrast, country which is in top 5 expenditure on higher education, their average score is not as good as the top scores ones.
    - Denmark takes place in both tables, this may means the government of Denmark in 2011 spent reasonably and their higher education system is also very good.
    - Looks like spending more on higher education does not have a relationship with getting a better average score.

##### An interesting fact about Denmark:

- According to the [Denmark spends most on education: OECD](https://www.thelocal.dk/20140909/denmark-spends-the-most-on-education-oecd-report/), A new OECD report reveals that Denmark has maintained its position as the country that invests the most in education (With 7.9 percent of its GDP in 2011 spent on public and private institutions at all levels of education) but unemployment rates for academics have doubled in recent years and teachers spend less time in the classroom than most other countries.